### ISSUE TITLE
The regular ROPs in boards EB2/EB4 saturates the UDP packet

### ISSUE THREAD

**Issue by marcoaccame** [robotology/icub-firmware#27]
In the boards EB2 and EB4 there are 12 joints/motors, skin and a MAIS whose status must be regularly signaled to robotInterface.

In these two boards, the available space reserved for regular ROPs is over and hence, there is no room for increasing size of the struct which define the status of joints and of motors (or of others).

However, there is now the need to revise and possibly increase these data structures. See minute of internal meeting: https://redmine.icub.iit.local/projects/firmware/wiki/ICubV3_-_technical_meetings'_reports_-_wiki_page_.

Thus, some solution is required, at least for the case of 12 joints, because for otehr boards with 4 joints or less the problem does not exist.

**Comment by marcoaccame** [comment]
One solution for boards EB2 / EB4  would be to alternate the transmission of the status of the joints/motors six at a time.

To send the complete status we would need the time of two UPD packets, but inside each packet there would be much more space to allow an easy increas of size of data structures.

Now the EB2 and EB4 boards transmits every 1 ms (and not 5 ms as the others) because of the limited size of skin.status. Thus if we maintain the 1 ms rate the complete status would be sent in just 2 ms which is still more than other boards.

This mechanism would not affect the other boards.

In more details:

```
case of EB2 / EB4: one frame every 1 ms
[regular-ropframe n+0] = [skin+mais] + [status jointsmotors from 0 to 5]
[regular-ropframe n+1] = [skin+mais] + [status jointsmotors from 6 to 11]
[regular-ropframe n+2] = [skin+mais] + [status jointsmotors from 0 to 5]
[regular-ropframe n+3] = [skin+mais] + [status jointsmotors from 6 to 11]


case of EB1 (for instance): 1 frame every 5 ms
[regular-ropframe n+0] = [strain] + [status jointsmotors from 0 to 3]
[regular-ropframe n+1] = [strain] + [status jointsmotors from 0 to 3]
[regular-ropframe n+2] = [strain] + [status jointsmotors from 0 to 3]
```

**Comment by marcoaccame** [comment]
The solution has been implemented and tested on the purple robot and it works

**Comment by marcoaccame** [comment]
committed sources (2250a7e7ed287633dda3f5ca4ffe16a59a470bfc and robotology/icub-firmware-shared@b9bc85b9640ae388d933be2f4cdae70ec2e54bc4) and binaries (robotology/icub-firmware-build@89d399806af39e6ba66a6fe54cf45348fe3cc497).

updated the purple robot.

**Comment by marcoaccame** [comment]
now in regular ropframe we have more room which enables us to increase the size of some data structures.

We have:
```
6*(sizeof(eOmc_joint_status_t) + sizeof(eOmc_motor_status_basic_t) = 
6*(36+16) = 
312 bytes available
```

With 312 bytes can now increase the status of a joint and/or motor up to 52 bytes. 

But maybe it is worth using some of 312 bytes for increasing the status of the skin (EOarray_of_skincandata_t). Now its size is 94 byte and contains 9 can frames and does not allow us to reduce the TX rate of the boards with skin.  

If we double to 18 can frames, we would use 90 bytes and we could use a tx rate of 2 ms or maybe also of 3 ms which makes the system more balanced.

With the remaining bytes we still could make the status of joint/motor fatter with 37 bytes.

**Comment by marcoaccame** [comment]
Preliminary tests on purple robot show that we can we can reduce the TX rate to 2 ms for boards with skin by using capacity equal 16 can frames.

And if we use capacity equal to 24 we can reduce to 3 ms.
In this second case, I have measured a good decrease of the CPU % in the PC104.

In comparison with the current situation, robotInterface goes from 41% to 34% and eth0 handler from 11% to 6%. 

Also, with capacity 24 we use 244 bytes for skin status instead of 96 of current case. Thus we have 312-(244-96)=164 bytes remaining for getting fatter the status of 6 joints/motors (27 bytes each).

I have also tested other cases: capacity equal 16 and 20. More details will be added later on.
